sdk/storage/faq.md

```diff
@@ -0,0 +1,374 @@
+
+// serviceClient sends HTTP requests with specified API-version.
+BlobClientOptions clientOptions;
+clientOptions.ApiVersion = "2019-12-12";
+auto serviceClient = BlobServiceClient::CreateFromConnectionString(GetConnectionString, clientOptions);
+// containerClient inherits this setting.
+auto containerClient = serviceClient.GetBlobContainerClient(containerName);
+```
+
+Please note that multiple-version is not supported yet by Storage C++ SDK (This may change in the future).
+Which means, the code above only changes the `x-ms-version` request header and doesn't change SDK behavior.
+SDK will still set headers and query parameters that are added between the customized API version (which is usually older) and latest API version,
+and still expect some headers or fields in HTTP response.
+Furthermore, this scenario is not covered by testing, although most of the APIs should be able to work. You should do thorough testing before your code goes to production.
+
+## How can I set a custom UserAgent string?
+
+We recommend you set an application ID with code below, so that it can be identified from which application, SDK, platform the request was sent. The information could be useful for troubleshooting and telemetry purposes.
+
+```C++
+BlobClientOptions clientOptions;
+clientOptions.Telemetry.ApplicationId = "SomeApplication v1.2.3";
+
+// The User-Agent string will be something like:
+// SomeApplication v1.2.3 azsdk-cpp-storage-blobs/12.9.0-beta.3 (Windows 10 Enterprise 6.3 19045 19041.1.amd64fre.vb_release.191206-1406)
+```
+
+## What encoding is used for input and output?
+
+All URLs should be URL-encoded, other resource names should be in UTF-8 encoding.
+This applies to both input variables and output.
+If your code runs in an environment where the default locale and encoding is not UTF-8, you should encode before passing variables into SDK and decode after reading a variable from SDK.
+
+```C++
+// containerUrl should be URL-encoded
+auto containerClient = BlobContainerClient(containerUrl);
+// The return value is URL-encoded
+auto url = containerClient.Geturl();
+// blobName should be UTF-8 encoded
+auto blobClient = containerClient.GetBlobClinet(blobName);
+
+for (auto page = blobContainerClient.ListBlobs(); page.HasPage(); page.MoveToNextPage()) {
+  for (auto& blob : page.Blobs) {
+    // blob.Name is UTF-8 encoded
+    std::cout << blob.Name << std::endl;
+  }
+}
+```
+
+## How can I check if a blob exists? Is there a function `Exists()` like in legacy SDK?
+
+There's no equivalent in the new C++ SDK. This function is usually just a convenient method on top of getting blob properties. You could easily write one like below.
+
+```C++
+bool BlobExists(const Azure::Storage::Blobs::BlobClient& client) {
+  try {
+    client.GetProperties();
+    return true;
+  }
+  catch (Azure::Storage::StorageException& e) {
+    if (e.ReasonPhrase == "The specified blob does not exist.") {
+      return false;
+    }
+    throw;
+  }
+}
```

Comment:
This is a bit of an anti-pattern (don't use exceptions as control flow). Is there a cleaner way to test for the existence of a blob?

https://isocpp.org/wiki/faq/exceptions#why-not-exceptions
> In particular, do not use exceptions for control flow. throw is not simply an alternative way of returning a value from a function (similar to return). Doing so will be slow and will confuse most C++ programmers who are rightly used to seeing exceptions used only for error handling.

Reply:
It isn't easy. It must be done at protocol layer, which is generated.

Tracked in https://github.com/Azure/azure-sdk-for-cpp/issues/5087

Reply:
What about this sample itself within the FAQ? We are encouraging an anti-pattern. Do we have another approach to implement `Exists()` on top of the current storage APIs?

Reply:
> What about this sample itself within the FAQ?

I'm afraid we have to live with it. There's no way to do this without catching exception.

If the problem of catching exception in normal flow is about performance penalty, I think it's not that horrible in this case, because we send requests over network which can take much longer time.